generators/chipyard/src/main/resources/csrc/cospike.cc

```diff
@@ -253,76 +283,121 @@ extern "C" void cospike_cosim(long long int cycle,
   if (valid) {
     printf("%d Cosim: %lx", cycle, iaddr);
     if (has_wdata) {
-      printf(" %lx", wdata);
+      printf(" s: %lx", wdata);
     }
     printf("\n");
   }
   if (valid || raise_interrupt || raise_exception) {
     p->step(1);
+#ifdef SPIKE_DEBUG
+    printf("spike pc is %lx\n", s->pc);
+    printf("spike mstatus is %lx\n", s->mstatus->read());
+    printf("spike mip is %lx\n", s->mip->read());
+    printf("spike mie is %lx\n", s->mie->read());
+#endif
   }
 
   if (valid) {
     if (s_pc != iaddr) {
-      printf("%d PC mismatch spike:%lx != dut:%lx\n", cycle, s_pc, iaddr);
+      printf("%d PC mismatch spike %llx != DUT %llx\n", cycle, s_pc, iaddr);
+#ifdef SPIKE_DEBUG
+      printf("spike mstatus is %lx\n", s->mstatus->read());
+      printf("spike mcause is %lx\n", s->mcause->read());
+      printf("spike mtval is %lx\n" , s->mtval->read());
+      printf("spike mtinst is %lx\n", s->mtinst->read());
+#endif
       exit(1);
     }
 
+
+  auto& mem_write = s->log_mem_write;
+  auto& log = s->log_reg_write;
+  auto& mem_read = s->log_mem_read;
+
+
+  for (auto memwrite : mem_write) {
+    reg_t waddr = std::get<0>(memwrite);
+    uint64_t w_data = std::get<1>(memwrite);
+    if (waddr == CLINT_BASE && w_data == 0) {
```

Comment:
I don't think CLINT_BASE is right, as it assumes hartid=0. Shouldn't it calculate the hartid based on address?